File: 백준/Platinum/17167. A Plus Equals B/A Plus Equals B.py
import math

V = []

sta = ["A+=A", "A+=B", "B+=A", "B+=B"]

A = 0
B = 0

def solve(A, B):

    while(A != B):
        if (A % 2 == 0 and B % 2 == 0):
            if ((A & (A - 1)) == 0 and (B & (B - 1)) == 0):
                if (A < B):
                    V.append([(int(math.log2(B / A))), 0])
                    A = B
                else:
                    V.append([(int(math.log2(A / B))), 3])
                    B = A
                continue

            T1 = A & (-A)
            T2 = B & (-B)
            c1 = 0 
            c2 = 0

            if (T1 > T2):
                B *= int(T1 / T2)
                V.append([int(math.log2(int(T1 / T2))), 3])
            
            elif (T1 < T2):
                A *= int(T2 / T1)
                V.append([int(math.log2(int(T2 / T1))), 0])

            c1 = int(A / max(T1, T2))
            c2 = int(B / max(T1, T2))

            if (c1 > c2):
                A += B
                V.append([1, 1])

            else:
                B += A
                V.append([1, 2])

        elif (A % 2 != 0 and B % 2 == 0):
            V.append([1, 0])
            A *= 2

        elif (A % 2 == 0 and B % 2 != 0):
            V.append([1, 3])
            B *= 2

        else:
            V.append([1, 0])
            V.append([1, 3])
            A *= 2
            B *= 2

A, B = map(int, input().split())

total = 0

solve(A, B)

for x, y in V:
    total += x

print(total)

for x, y in V:
    for j in range(x):
        print(sta[y])

File: 백준/Platinum/17167. A Plus Equals B/A Plus Equals B.cc
#include <bits/stdc++.h>
#define fastio ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
#define ll long long
#define o (ll)1

using namespace std;

ll A, B;

string sta[]{ "A+=A", "A+=B", "B+=A", "B+=B" };

vector<ll> V;

void solve()
{
	while (A != B)
	{
		ll T1 = A & (-A);
		ll T2 = B & (-B);

		if (T1 > T2)
		{
			for (ll i = 0; i < (ll)(log2(T1 / T2)); i++)
				V.push_back(3);
		}
		else if (T1 < T2)
		{
			for (ll i = 0; i < (ll)(log2(T2 / T1)); i++)
				V.push_back(0);
		}
		
		A >>= (ll)(log2(T1)), B >>= (ll)(log2(T2));
	
		if (A < B)
		{
			B += A;
			V.push_back(2);
		}
		else if (A > B)
		{
			A += B;
			V.push_back(1);
		}
	}
}

int main()
{
	fastio;

	cin >> A >> B;

	solve();

	cout << (ll)V.size() << '\n';
	
	for (auto e : V)
		cout << sta[e] << '\n';
}